x-pack/test/functional/services/ml/swim_lane.ts

```diff
@@ -0,0 +1,215 @@
+  const testSubjects = getService('testSubjects');
+
+  /**
+   * Y axis labels width + padding
+   */
+  const xOffset = 185;
+
+  /**
+   * Get coordinates relative to the left top corner of the canvas
+   * and transpose them from the center point.
+   */
+  async function getCoordinatesFromCenter(
+    el: WebElementWrapper,
+    coordinates: { x: number; y: number }
+  ) {
+    const { width, height } = await el.getSize();
+
+    const elCenter = {
+      x: Math.round(width / 2),
+      y: Math.round(height / 2),
+    };
+
+    /**
+     * Origin of the element uses the center point, hence we need ot adjust
+     * the click coordinated accordingly.
+     */
+    const resultX = xOffset + Math.round(coordinates.x) - elCenter.x;
+    const resultY = Math.round(coordinates.y) - elCenter.y;
+
+    return {
+      x: resultX,
+      y: resultY,
+    };
+  }
+
+  const getRenderTracker = async (testSubj: string) => {
+    const renderCount = await elasticChart.getVisualizationRenderingCount(testSubj);
+
+    return {
+      async verify() {
+        if (testSubj === 'mlAnomalyExplorerSwimlaneViewBy') {
+          // We have a glitchy behaviour when clicking on the View By swim lane.
+          // The entire charts is re-rendered, hence it requires a different check
+          await testSubjects.existOrFail(testSubj);
+          await elasticChart.waitForRenderComplete(testSubj);
+        } else {
+          await elasticChart.waitForRenderingCount(renderCount + 1, testSubj);
+        }
+      },
+    };
+  };
+
+  return {
+    async getDebugState(testSubj: string): Promise<HeatmapDebugState> {
+      const state = await elasticChart.getChartDebugData(testSubj);
+      if (!state) {
+        throw new Error('Swim lane debug state is not available');
+      }
+      return state as HeatmapDebugState;
+    },
+
+    async getAxisLabels(testSubj: string, axis: 'x' | 'y'): Promise<DebugStateAxis['labels']> {
+      const state = await this.getDebugState(testSubj);
+      return state.axes[axis][0].labels;
+    },
+
+    async assertAxisLabels(testSubj: string, axis: 'x' | 'y', expectedValues: string[]) {
+      const actualValues = await this.getAxisLabels(testSubj, axis);
+      expect(actualValues).to.eql(
+        expectedValues,
+        `Expected swim lane ${axis} labels to be ${expectedValues}, got ${actualValues}`
+      );
+    },
+
+    async getCells(testSubj: string): Promise<HeatmapDebugState['heatmap']['cells']> {
+      const state = await this.getDebugState(testSubj);
+      return state.heatmap.cells;
+    },
+
+    async getHighlighted(testSubj: string): Promise<HeatmapDebugState['heatmap']['selection']> {
+      const state = await this.getDebugState(testSubj);
+      return state.heatmap.selection;
+    },
+
+    async assertSelection(
+      testSubj: string,
+      expectedData: HeatmapDebugState['heatmap']['selection']['data'],
+      expectedArea?: HeatmapDebugState['heatmap']['selection']['area']
```

Comment:
It looks like `expectedArea` is not part of the assertion, but it should probably be?

Reply:
I decided to keep it optional because this data is not important for existing tests

Reply:
Not sure I understand this correctly. So you've added this optional parameter because it might be used in the future, but is not used today, right? If that's the case, wouldn't it be better to not add the parameter yet but add a TODO comment? Or is there a specific reason why this parameter should be there although it's not implemented (yet)?

Currently, a user of the service sees something like this:
![image](https://user-images.githubusercontent.com/1945390/112304651-0619af00-8c9e-11eb-9fbc-7cd6d3826bf0.png)
and might be under the impression that  a passed `expectedArea` would be validated in some way :thinking:

Reply:
ah yes, I forgot something, check https://github.com/elastic/kibana/pull/94723/commits